lib/codecs/src/decoding/framing/chunked_gelf.rs

```diff
@@ -0,0 +1,735 @@
+        if self.is_complete() {
+            self.timeout_task.abort();
+            let chunks = &self.chunks[0..self.total_chunks as usize];
+            let mut message = BytesMut::new();
+            for chunk in chunks {
+                message.extend_from_slice(chunk);
+            }
+            Some(message.freeze())
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Debug, Snafu)]
+pub enum ChunkedGelfDecoderError {
+    #[snafu(display("Invalid chunk header with less than 10 bytes: {header:?}"))]
+    InvalidChunkHeader { header: Bytes },
+    #[snafu(display("Invalid total chunks value {total_chunks} for message with id {message_id} and sequence number {sequence_number}"))]
+    InvalidTotalChunks {
+        message_id: u64,
+        sequence_number: u8,
+        total_chunks: u8,
+    },
+    #[snafu(display("Sequence number {sequence_number} is greater than the total chunks value {total_chunks} for message with id {message_id}"))]
+    InvalidSequenceNumber {
+        message_id: u64,
+        sequence_number: u8,
+        total_chunks: u8,
+    },
+    #[snafu(display("Pending messages limit of {pending_messages_limit} reached while processing message with id {message_id} and sequence number {sequence_number}"))]
+    PendingMessagesLimitReached {
+        message_id: u64,
+        sequence_number: u8,
+        pending_messages_limit: usize,
+    },
+    #[snafu(display("Received message with id {message_id} and sequence number {sequence_number} has different total chunks values: original total chunks value {original_total_chunks}, received total chunks value {received_total_chunks}"))]
+    TotalChunksMismatch {
+        message_id: u64,
+        sequence_number: u8,
+        original_total_chunks: u8,
+        received_total_chunks: u8,
+    },
+}
+
+impl StreamDecodingError for ChunkedGelfDecoderError {
+    fn can_continue(&self) -> bool {
+        true
+    }
+}
+
+impl FramingError for ChunkedGelfDecoderError {
+    fn as_any(&self) -> &dyn Any {
+        self as &dyn Any
+    }
+}
+
+/// A codec for handling GELF messages that may be chunked. The implementation is based on [Graylog's GELF documentation](https://go2docs.graylog.org/5-0/getting_in_log_data/gelf.html#GELFviaUDP)
+/// and [Graylog's go-gelf library](https://github.com/Graylog2/go-gelf/blob/v1/gelf/reader.go).
+#[derive(Debug, Clone)]
+pub struct ChunkedGelfDecoder {
+    // We have to use this decoder to read all the bytes from the buffer first and don't let tokio
+    // read it buffered, as tokio FramedRead will not always call the decode method with the
+    // whole message. (see https://docs.rs/tokio-util/latest/src/tokio_util/codec/framed_impl.rs.html#26).
+    // This limitation is due to the fact that the GELF format does not specify the length of the
+    // message, so we have to read all the bytes from the message (datagram)
+    bytes_decoder: BytesDecoder,
+    state: Arc<Mutex<HashMap<u64, MessageState>>>,
```

Comment:
This would be a performance optimization, but we could think about using `dashmap` here to reduce lock contention. As it stands, it seems like only one packet can be processed at a time due to the locking, which may limit throughput.

Reply:
I think the udp socket source does not parallelize and the whole process of reading from the socket -> framing -> decoding is entirely sequential from the udp socket's perspective.

I may be wrong but I think https://github.com/vectordotdev/vector/blob/17466c684bc6d3b0356c976e9639f9b05d306769/src/sources/socket/udp.rs#L177 is purely sequential. 

So I think the lock would only compete with the message timeout acquires, but in that case, if there are a lot of concurrent timeouts, you are right that dashmap couls help

Anyway, I think it would be useful for this last case, but I'd like if you can confirm the first part about the socket's sequential reads

Reply:
I implemented it with dashmap in https://github.com/jorgehermo9/vector/commit/3e5f57f027ca21709b44758db0b235d527a558a8 at another branch (https://github.com/jorgehermo9/vector/tree/feature/chunked-gelf-dashmap) and it seems that my implementation with it causes deadlocks and several tests fails
![image](https://github.com/user-attachments/assets/b63e7725-b543-4d3f-af93-5a015ab9f4d7)

I'm concerned about the usage of the [entry](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.entry) and [remove](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html#method.remove) methods, which seems to be the cause of the deadlocks. The `message_state` reference that is returned by the `entry` method lives until the end of the scope, but we need to call `remove` in some branches before returning in the `decode_chunk` method, so I think thats the cause of the deadlock. 

I don't know how to model what I'm doing with `DashMap`

Reply:
Mmm, yeah, you are right that for UDP packets are read and processed sequentially at the moment though would like to improve that in the future (https://github.com/vectordotdev/vector/issues/8518) so this may turn into a bottleneck at that point.

I see for the TCP sockets it uses one decoder instance per stream, same for Unix stream sockets, and for Unix datagram sockets it is similar to UDP where all packets are processed sequentially.

We can go without this for the initial implementation and I can play around with trying to swap in a dashmap after it is merged.

Reply:
In UDP there is no concept of multiple streams as in TCP, so I think having multiple decoder instances is not worth.

The alternative in this case is to use a worker per message, but I don't think it would be worth as there would be *a lot* of workers doing relatively small work. Maybe we could have a pool of workers (for example, 10, configurable) and use [crossbeam channel](https://docs.rs/crossbeam/latest/crossbeam/channel/index.html) to decouple the socket reads from the decoder's work.

I may give a try to it if we found performance issues the UDP socket (If we add decompression for the GELF chunks, I think this could really be a bottleneck)

Reply:
> In UDP there is no concept of multiple streams as in TCP

That's true, but we did discuss in https://github.com/vectordotdev/vector/issues/8518 the idea of creating a separate "stream" by grouping together packets from the same source IP / port.

Definitely no need to do this now, but if you are ever interested, https://github.com/DataDog/lading is the tool we use to generate load for benchmarks. We could add the option to generate GELF data there in order to add a a [benchmark experiment to Vector](https://github.com/vectordotdev/vector/tree/master/regression/cases).